content/blog/2022/2022-02-14-spring-shell.md

```diff
@@ -0,0 +1,331 @@
+    private boolean signedIn;    
+    
+    @ShellMethod(value = "sign in")    
+    public void signIn()    
+    {    
+        this.signedIn = true;    
+        log.info("Signed In!");    
+    }    
+    
+    @ShellMethod(value = "sign out")    
+    public void signOut()    
+    {    
+        this.signedIn = false;    
+        log.info("Signed out!");    
+    }    
+    // note the naming     
+   public Availability signOutAvailability()    
+    {    
+        return signedIn ?    
+                Availability.available() : Availability.unavailable("Must be signed in first");    
+	} 
+}  
+``` 
+So if we try to run the `sign-out` command without first signing in we will get the following message:
+```bash  
+shell:>sign-out  
+Command 'sign-out' exists but is not currently available because Must be signed in firstDetails of the error have been omitted. You can use the stacktrace command to print the full stacktrace.shell:>  
+```  
+
+### The Second Way
+Uses the `@ShellMethodAvailability` annotation, in which we specify the method name we want to use to Availability check:
+```java  
+@ShellComponent  
+public  class  SSHLoggingCommand  
+{
+	...
+	@ShellMethod(value = "sign out") 
+	@ShellMethodAvailability("signOutCheck") 
+	public void signOut() 
+	{    
+	    this.signedIn = false;    
+		log.info("Signed out!"); 
+	}  
+		
+	public Availability signOutCheck() 
+	{    
+	   return signedIn ?  Availability.available() : Availability.unavailable("Must be signed in first"); 
+	}  
+}
+``` 
+
+### The Third Way
+It enables us to have several methods attached to a single availability method.  
+We are going to use the annotation `ShellMethodAvailability` with an array of the **commands names (not method names)**:
+
+```java  
+@ShellComponent  
+public  class  SSHLoggingCommand  
+{
+	...
+	@ShellMethod(value = "sign out") public void signOut() 
+	{    
+	    this.signedIn = false;    
+		log.info("Signed out!"); 
+	}    
+
+	 @ShellMethod(value = "Change password") 
+	 public void changePass(@ShellOption String newPass) 
+	 {    
+		log.info(format("Changed password to '%s'", newPass)); 
+	 }    
+	 
+	 @ShellMethodAvailability({"sign-out", "change-pass"}) 
+	 public Availability signOutCheck() 
+	 {    
+	    return signedIn ? Availability.available() : Availability.unavailable("Must be signed in first"); 
+	}  
+}
+```  
+
+## Other Cool Features in Spring Shell
+Since Spring Shell builds on top of [JLine](https://github.com/jline/jline3) it inherits a lot of its features. Let's look at some of them:
+
+### Tab Completion
+Spring Shell allows us to use tab completion with command names and even with parameter names.
+
+### Built in Commands
+
+- help:
+  - lists all the commands known to the shell, including the built-in commands and commands we wrote.
+
+- script:
+  - accepts a local file as an argument and will replay commands found there, one at a time.
```

Comment:
Please make this real text instead of just an enumeration. Or add an intro sentence, then enumerate the commands (but don't use a two-level list). I.e. something like this:

Intro text

- **`help`:** lists all...
- **`script`:** accepts a ...